File: app/desktop/src/lib/types/physics.ts
/**
 * Simulation result contracts: force sweep, stackup / power / friction
 * budgets, and the B-field sampling grid.
 */

import type { CommutationMode } from "./enums";

/** Force-vs-position sweep (evaluate_force_sweep). */
export interface ForceSweepResult {
  positions_m: number[];
  force_x_n: number[];
  force_y_n: number[];
  force_z_n: number[];
  per_phase_force_x: number[][];
  commutation: CommutationMode;
  current_a: number;
  mean_thrust_n: number;
  peak_thrust_n: number;
  min_thrust_n: number;
  ripple_pct: number;
  n_positions: number;
}

/** Layer stackup summary (compute_stackup). */
export interface StackupResultDto {
  layer_count: number;
  trace_widths_m: number[];
  cu_thickness_m: number[];
  via_drill_m: number;
  via_annular_ring_m: number;
  via_grid_rows: number;
  via_grid_cols: number;
  estimated_force_n: number;
  estimated_dc_resistance_ohm: number;
  notes: string[];
}

/** Height stack cross-section (compute_height_stack). */
export interface HeightStackResultDto {
  pcb_thickness_m: number;
  cu_protrusion_m: number;
  solder_mask_m: number;
  air_gap_m: number;
  magnet_height_m: number;
  tolerance_m: number;
  total_height_m: number;
}

/** Friction budget breakdown (compute_friction). */
export interface FrictionBudgetDto {
  bearing_friction_n: number;
  cable_drag_n: number;
  wiper_contact_n: number;
  cogging_n: number;
  total_n: number;
  minimum_drive_force_n: number;
}

/** Power / thermal budget (compute_power_budget). */
export interface PowerBudgetDto {
  phase_resistance_ohm: number;
  continuous_power_w: number;
  burst_power_w: number;
  temperature_rise_c: number;
  capacitor_required_uf: number;
  efficiency_pct: number;
}

/** One B-field sample on the X–Z sampling grid (sample_b_field). */
export interface BFieldSampleDto {
  x_m: number;
  z_m: number;
  bx_t: number;
  by_t: number;
  bz_t: number;
  mag_t: number;
}

/** B-field grid (sample_b_field). */
export interface BFieldGridDto {
  samples: BFieldSampleDto[];
  /** [x_min, x_max] [m] */
  x_extent_m: [number, number];
  /** [z_min, z_max] [m] */
  z_extent_m: [number, number];
}

File: app/desktop/src-tauri/src/ipc/physics.rs
//! IPC physics-result DTOs: B-field grid, force sweep, stackup, height stack,
//! friction budget, and power budget.

use serde::{Deserialize, Serialize};

use super::enums::CommutationModeIpc;

// ===========================================================================
// B-field grid (sample_b_field) — WP4 / WP5 flux-viz backend
// ===========================================================================

/// One B-field sample on the X–Z flux-viz grid.
///
/// All units SI: positions in metres, B-field in Tesla. `mag_t` is the
/// precomputed `sqrt(bx² + by² + bz²)` — the Svelte renderer uses it to
/// colour-code arrows by magnitude without recomputing.
///
/// Field naming on the wire: every field has an explicit `#[serde(rename)]`
/// so the unit suffix (`_m`, `_t`) is preserved end-to-end and stable
/// across Rust → JSON → TypeScript.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BFieldSampleIpc {
    #[serde(rename = "x_m")]
    pub x_m: f64,
    #[serde(rename = "z_m")]
    pub z_m: f64,
    #[serde(rename = "bx_t")]
    pub bx_t: f64,
    #[serde(rename = "by_t")]
    pub by_t: f64,
    #[serde(rename = "bz_t")]
    pub bz_t: f64,
    #[serde(rename = "mag_t")]
    pub mag_t: f64,
}

impl BFieldSampleIpc {
    /// Convert a core `pcbmotorgen_simulation::magnetic::BFieldSample2D` to the
    /// IPC wire form, computing the magnitude on the way out.
    pub fn from_core(s: &pcbmotorgen_simulation::magnetic::BFieldSample2D) -> Self {
        Self {
            x_m: s.x,
            z_m: s.z,
            bx_t: s.bx,
            by_t: s.by,
            bz_t: s.bz,
            mag_t: (s.bx * s.bx + s.by * s.by + s.bz * s.bz).sqrt(),
        }
    }
}

/// Full 2D B-field grid response for the `sample_b_field` Tauri command.
///
/// `samples` is **row-major** with Z as the slow axis:
/// `samples[i_z * n_x + i_x]`. The Svelte `FluxDiagram` reshapes the flat
/// `samples` into a 2D `n_z × n_x` arrow grid using `x_extent_m` /
/// `z_extent_m` to recover the physical axes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct BFieldGridIpc {
    pub samples: Vec<BFieldSampleIpc>,
    /// `[x_min, x_max]` over which the grid was sampled [m].
    pub x_extent_m: [f64; 2],
    /// `[z_min, z_max]` over which the grid was sampled [m].
    pub z_extent_m: [f64; 2],
}

// ===========================================================================
// Force sweep (evaluate_force_sweep)
// ===========================================================================

/// Force vs mover position along the travel axis.
///
/// Per PRODUCT_GOALS §4.C the sign convention is `F_mover = -F_stator`;
/// `force_x_n` already reflects the mover's reference frame.
/// Ripple % = `(F_max - F_min) / |F_mean| × 100` (§4.A).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ForceSweepResultIpc {
    pub positions_m: Vec<f64>,
    pub force_x_n: Vec<f64>,
    pub force_y_n: Vec<f64>,
    pub force_z_n: Vec<f64>,
    /// Per-phase x-force at each position: `[position][phase]`.
    pub per_phase_force_x: Vec<Vec<f64>>,
    pub commutation: CommutationModeIpc,
    pub current_a: f64,
    pub mean_thrust_n: f64,
    pub peak_thrust_n: f64,
    pub min_thrust_n: f64,
    pub ripple_pct: f64,
    pub n_positions: u32,
}

// ===========================================================================
// Stackup / height / power / friction (compute_*)
// ===========================================================================

/// PCB stackup recommendation (trace widths, copper, vias).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StackupResultIpc {
    pub layer_count: u32,
    pub trace_widths_m: Vec<f64>,
    pub cu_thickness_m: Vec<f64>,
    pub via_drill_m: f64,
    pub via_annular_ring_m: f64,
    pub via_grid_rows: u32,
    pub via_grid_cols: u32,
    pub estimated_force_n: f64,
    pub estimated_dc_resistance_ohm: f64,
    pub notes: Vec<String>,
}

/// Explicit vertical stack from PCB bottom to magnet top.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct HeightStackResultIpc {
    pub pcb_thickness_m: f64,
    pub cu_protrusion_m: f64,
    pub solder_mask_m: f64,
    pub air_gap_m: f64,
    pub magnet_height_m: f64,
    pub tolerance_m: f64,
    pub total_height_m: f64,
}

/// Mechanical friction breakdown.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct FrictionBudgetIpc {
    pub bearing_friction_n: f64,
    pub cable_drag_n: f64,
    pub wiper_contact_n: f64,
    /// Coreless motor → zero cogging (PRODUCT_GOALS §4.A).
    pub cogging_n: f64,
    pub total_n: f64,
    pub minimum_drive_force_n: f64,
}

/// Stable-equilibrium travel envelope of the mover array centre: the
/// charge-based electromagnetic endpoints (kata k5r5) clamped into the
/// span-aware flush limits (kata 5c7r). Every stable rest centre satisfies
/// `rest_phase_m ≡ x (mod electrical_period_m)`; the slider range endpoints
/// are the travel limits, not the rests.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct TravelEnvelopeIpc {
    /// First stable rest position of the array centre [m].
    pub min_position_m: f64,
    /// Last stable rest position of the array centre [m] (≥ min).
    pub max_position_m: f64,
    /// Rest phase φ [m].
    pub rest_phase_m: f64,
    /// Electrical period P_e = 2 × pole pitch (one full 360° electrical cycle) [m].
    pub electrical_period_m: f64,
}

impl From<pcbmotorgen_simulation::equilibrium::TravelEnvelope> for TravelEnvelopeIpc {
    fn from(env: pcbmotorgen_simulation::equilibrium::TravelEnvelope) -> Self {
        Self {
            min_position_m: env.min_position_m,
            max_position_m: env.max_position_m,
            rest_phase_m: env.rest_phase_m,
            electrical_period_m: env.electrical_period_m,
        }
    }
}

/// Continuous and burst power / thermal analysis.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct PowerBudgetIpc {
    pub phase_resistance_ohm: f64,
    pub continuous_power_w: f64,
    pub burst_power_w: f64,
    pub temperature_rise_c: f64,
    pub capacitor_required_uf: f64,
    pub efficiency_pct: f64,
}